xapian-core/weight/pl2weight.cc

```diff
@@ -78,13 +78,52 @@ PL2Weight::init(double)
 
     // Calculate the lower bound on the weight.
     double P_min =
-	P1 + (wdfn_lower + 0.5) * log2(wdfn_lower) - P2 * wdfn_lower;
+	P1 + (wdfn_lower + 0.5) * log2(wdfn_lower) - P2 * wdfn_upper;
     lower_bound = get_wqf() * P_min / (wdfn_upper + 1.0);
 
     // Calculate the upper bound on the weight.
-    double P_max =
-	P1 + (wdfn_upper + 0.5) * log2(wdfn_upper) - P2 * wdfn_upper;
-    upper_bound = get_wqf() * P_max / (wdfn_lower + 1.0);
+    double term2;
+
+    // P = ((wdfn + 0.5) * log2(wdfn) + P1 - (P2 * wdfn)) / (wdfn + 1.0).
+    // We break the term into three parts with (wdfn + 1.0) as the denominator
+    // in each term.
+    // The maximum of the first term, i.e (wdfn + 0.5) * log2(wdfn) /
+    // (wdfn + 1.0), will happen at wdfn_upper as it is an increasing function.
+    // This can be verified by differentiating it and observing that
+    // the differentiation always has a positive value.
+    // The third term, i.e -P2 * wdfn / (wdfn + 1.0), needs to be minimized as it
+    // is negative.
+    // The third term will be minimum at wdfn_lower as it is an increasing
+    // function.
+    // In order to determine the nature of the second term, i.e P1 /
+    // (wdfn + 1.0) in colloboration with the first term, we differentiate the
+    // expression of the first and the second terms i.e (P1 + (wdfn  + 0.5) *
+    // log2(wdfn)) / (wdfn + 1.0).
+    // The numerator of the differentiation can be negative.
+    // So, we again differentiate the numerator and check the value of the
+    // second and the first differentiation at wdfn_upper.
```

Comment:
While the details here aren't quite right, the idea of splitting the equation up and maximising each part individually is certainly sound.

The best approach seems to be to split it into two parts - one with the `log` and one for the rest.  When `P1 + P2 > 0` (which is often true) then both are optimised at `wdfn_upper`.

I've implemented that in 0964391357ec249cf133c288685e7fab63425eee.

This makes a huge difference in the upper bound - in a simple test, the upper bounds are now just over double the highest weight actually achieved - previously they were several hundred times.  So many thanks for coming up with this approach.